feat(cli): add ado show trace command (#1035)
IBM/ado: website/docs/getting-started/ado.md

ado show results operation randomwalk-0.5.0-123abc --output-file results-table.txt
```

#### ado show trace

> _show trace_ allows inspecting in detail the trace of requests made during

Reviewer: ```suggestion
_show trace_ allows inspecting in detail the trace of entity measurement requests made during
```

Author: 28f25e957660da380fd9965294ad8da1779f195a

explore operations. It can provide crucial information for debugging operation

> behaviour e.g. failed experiments or requests.

Reviewer: ```suggestion
behaviour e.g. failed experiments or requests. Note, multiple entities can be contained in a single measurement request depending on the sampler used to explore and its settings. 
```

Author: 28f25e957660da380fd9965294ad8da1779f195a

The complete syntax of the `ado show trace` command is as follows:

<!-- markdownlint-disable line-length -->

```shell
ado show trace operation [RESOURCE_ID] [--use-latest] \
                         [--include-results] \
                         [--filter <key=value>] \
                         [--output | -o <table | csv | json | yaml>] \
                         [--output-file <path>] \
                         [--hide <field>] \
                         [--no-trunc]
```

<!-- markdownlint-enable line-length -->

- `--use-latest` will use the identifier of the latest (i.e. most recent)
  operation from the current context. It is ignored if a RESOURCE_ID is
  provided.
- `--include-results` switches to result-level view, unrolling entities to show

>   individual measurement results with both request and result metadata.

Reviewer: This sounds like it includes actual results but it doesn't. Maybe it should be 
```suggestion
- `--include-results-metadata`  expands the table for output mode `table` or `csv` so each entity has its own row containing additional metadata on the result of applying the requested experiment to it
```

Also its only relevant for the table output types

Author: I went with --unroll-entities which is shorter and I think still conveys the correct meaning - lmk what you think.

f22b5065339457e2175a415ba6e11099cc1e1e67

- `--filter` filters using JSON path syntax based on **YAML field names** from
  the underlying data model, not table column names (e.g., `requestIndex=5`,
  `status=Success`). Can be used multiple times for AND logic. Only

>   request-level filters are supported.

Reviewer: Since we can't actually filter on all the fields in the YAML model I think we can leave out details like, json path syntax, YAML names, request-level filters

```suggestion
- `--filter` filters based on field names from the underlying data model, not
  table column names. Can be used multiple times for AND logic.
```

Author: 28f25e957660da380fd9965294ad8da1779f195a

  **Available filter fields:**

  Request-level fields (from MeasurementRequest):

  - `requestIndex` - Request index number
  - `requestid` - Request UUID
  - `status` - Request status (Unknown, Success, Failed)
  - `timestamp` - Request timestamp
  - `metadata` - Request metadata (use dot notation for nested fields, e.g.,
    `metadata.key=value`)

>   - `experimentReference` - Stringified experiment reference

Reviewer: I suggest moving this after the Expanded Trace Output Table and change it so 
```suggestion
  #### Filtering Output

The output of show trace can be filtered using the following fields:

  - `requestIndex` - Request index number
  - `requestid` - Request UUID
  - `status` - Request status (Unknown, Success, Failed)
  - `timestamp` - Request timestamp
  - `metadata` - Request metadata (use dot notation for nested fields, e.g.,
    `metadata.key=value`)
  - `experimentReference` - Stringified experiment reference
  
Filtering reduces the output to the requests matching the filters.   

```

Author: 28f25e957660da380fd9965294ad8da1779f195a

- `--output` (or `-o`) determines the output format. Supports `table`, `csv`,
  `json`, and `yaml`. Output is written to stdout by default, or to a file if
  `--output-file` is specified.
- `--output-file` specifies a file path to write the output to. If not provided,
  output is written to stdout.
- `--hide` can be specified multiple times and allows hiding fields from the
  output.
- `--no-trunc` prevents truncation of table content (console output only).

##### Default Trace Output Table

The default output table shows the time-series of measurement requests with the
following columns:

- Index (auto-generated row number)
- Request ID
- Request Index
- Request type (measured/replayed)
- Timestamp
- Experiment ID
- Entity IDs (list)
- Status
- Measurements (count)
- Valid Measurements (count)
- Invalid Measurements (count)
- Metadata (request metadata)

##### Expanded Trace Output Table

Specifying `--include-results` unrolls each request so each entity with a

> request processed has its own row with greater detail:

Reviewer: ```suggestion
unrolls each request so each entity with a
request processed has its own row with greater detail:

Specifying `--include-results-metadata ` expands the table so each entity has its own row containing additional metadata on the result of applying the requested experiment to it.
```

Author: I went with --unroll-entities which is shorter and I think still conveys the correct meaning - lmk what you think.

f22b5065339457e2175a415ba6e11099cc1e1e67

- Timestamp
- Experiment ID
- Entity IDs (list)
- Status
- Measurements (count)
- Valid Measurements (count)
- Invalid Measurements (count)
- Metadata (request metadata)

##### Expanded Trace Output Table

Specifying `--unroll-entities` unrolls each request so each entity with a
request processed has its own row containing additional metadata on the result
of applying the requested experiment to it:

- Index (auto-generated row number)
- Request ID
- Request Index
- Request type (measured/replayed)
- Timestamp
- Experiment ID
- Result Index (per-request, 0-based)
- Result UID
- Entity ID (single, unrolled)
- Valid (boolean)
- Number of Properties
- Request Metadata
- Result Metadata

#### Filtering Output

The output of show trace can be filtered using the following fields:

- `requestIndex` - Request index number
- `requestid` - Request UUID
- `status` - Request status (Unknown, Success, Failed)
- `timestamp` - Request timestamp
- `metadata` - Request metadata (use dot notation for nested fields, e.g.,
  `metadata.key=value`)
- `experimentReference` - Stringified experiment reference

Filtering reduces the output to the requests matching the filters.

##### Examples

> ###### Show the trace for an operation

Reviewer: ```suggestion
###### Show the trace for an operation as a table
```

```shell
ado show trace operation randomwalk-0.5.0-123abc
```

> ###### Show result level information in the trace

Reviewer: ```suggestion
###### Show entity level information in the trace table
```

- Valid Measurements (count)
- Invalid Measurements (count)
- Metadata (request metadata)

##### Result-Level View (--include-results)

The result-level view unrolls entities to show individual results:

- Index (auto-generated row number)
- Request ID
- Request Index
- Request type (measured/replayed)
- Timestamp
- Experiment ID
- Result Index (per-request, 0-based)
- Result UID
- Entity ID (single, unrolled)
- Valid (boolean)
- Number of Properties
- Request Metadata
- Result Metadata

##### Examples

###### Show request-level trace for an operation

```shell
ado show trace operation randomwalk-0.5.0-123abc
```

###### Show result-level trace with unrolled entities

```shell
ado show trace operation randomwalk-0.5.0-123abc --include-results
```

###### Filter by request status

```shell
ado show trace operation randomwalk-0.5.0-123abc --filter status=Success
```

###### Filter by request index

```shell
ado show trace operation randomwalk-0.5.0-123abc --filter requestIndex=5
```

###### Multiple filters (AND logic)

<!-- markdownlint-disable line-length -->

```shell
ado show trace operation randomwalk-0.5.0-123abc --filter status=Success --filter requestIndex=5

> ```

Reviewer: I suggest adding a list of fields that can be filtered on. Or e.g. giving an example YAML and saying any of these fields can be filtered on via json path?

Then for example I think we can simplify to just key ones and rely on convention and the CLI docs for people to work out rest? 

Key examples
- Getting a single measurement request YAML
- Filtering non successful requests, or requests with InvalidMeasurementResults. 

I'd hope if we establish early in the docs that -o can be used to output in different formats, that we won't have to state -o options for every command ...

ado show results operation randomwalk-0.5.0-123abc --output-file results-table.txt
```

#### ado show trace

_show trace_ provides a detailed view of measurement requests and results for
debugging purposes. For normal data analysis and visualization, use
`ado show measurements` instead, which provides a cleaner view of the measured

> data.

Reviewer: ```suggestion
_show trace_ allows inspecting in detail the trace of requests made during
explore operations. It can provide crucial information for debugging operation
behaviour e.g. failed experiments or requests.
```

  - `experimentReference.actuatorIdentifier` - Actuator ID
  - `entities[N].identifier` - Entity identifier at index N

  Result-level fields (from MeasurementResult, requires `--include-results` or
  auto-enabled):

  - `measurements[N].uid` - Result UUID at index N
  - `measurements[N].entityIdentifier` - Entity identifier for result at index N
  - `measurements[N].metadata` - Result metadata (use dot notation for nested
    fields)
  - `measurements[N].experimentReference.experimentIdentifier` - Experiment ID
    for result
  - `measurements[N].reason` - Failure reason (for invalid results)
  - `measurements[N].measurements[M].property.identifier` - Property identifier
    (for valid results)

- `--output` (or `-o`) determines the output format. Supports `table`, `csv`,
  `json`, and `yaml`. Output is written to stdout by default, or to a file if
  `--output-file` is specified.
- `--output-file` specifies a file path to write the output to. If not provided,
  output is written to stdout.
- `--hide` can be specified multiple times and allows hiding fields from the
  output.
- `--no-trunc` prevents truncation of table content (console output only).

> ##### Request-Level View (Default)

Reviewer: ```suggestion
##### Default Trace Output Table
```

> The default view shows measurement requests with the following columns:

Reviewer: ```suggestion
The default output table shows the time-series of measurement requests with the
following columns:
```

- Index (auto-generated row number)
- Request ID
- Request Index
- Request type (measured/replayed)
- Timestamp
- Experiment ID
- Entity IDs (list)
- Status
- Measurements (count)
- Valid Measurements (count)
- Invalid Measurements (count)
- Metadata (request metadata)

> ##### Result-Level View (--include-results)

Reviewer: ```suggestion
##### Expanded Trace Output Table
```

> The result-level view unrolls entities to show individual results:

Reviewer: ```suggestion
Specifying `--include-results` unrolls each request so each entity with a
request processed has its own row with greater detail:
```

- Index (auto-generated row number)
- Request ID
- Request Index
- Request type (measured/replayed)
- Timestamp
- Experiment ID
- Result Index (per-request, 0-based)
- Result UID
- Entity ID (single, unrolled)
- Valid (boolean)
- Number of Properties
- Request Metadata
- Result Metadata

##### Examples

> ###### Show request-level trace for an operation

Reviewer: ```suggestion
###### Show the  trace for an operation
```

```shell
ado show trace operation randomwalk-0.5.0-123abc
```

> ###### Show result-level trace with unrolled entities

Reviewer: ```suggestion
###### Show result level information in the trace
```

```shell
ado show trace operation randomwalk-0.5.0-123abc --include-results
```

> ###### Multiple filters with AND logic (YAML fields)

Reviewer: ```suggestion
###### Filter trace on multiple request fields
```

<!-- markdownlint-disable line-length -->

```shell
ado show trace operation randomwalk-0.5.0-123abc --filter status=Success --filter requestIndex=5
```

<!-- markdownlint-enable line-length -->

###### Filter by result UID (auto-enables --include-results)

<!-- markdownlint-disable line-length -->

```shell
ado show trace operation randomwalk-0.5.0-123abc --filter 'measurements[0].uid=result-uuid'

> ```

Reviewer: We miss the most crucial example which is getting the request YAML
